src/main/scala/sigmastate/utxo/transformers.scala

```diff
@@ -247,23 +247,22 @@ object Fold {
     Fold[SByteArray](input, 21, ByteArrayConstant(Array.emptyByteArray), 22, AppendBytes(TaggedByteArray(22), TaggedByteArray(21)))
 }
 
-case class ByIndex[V <: SType](input: Value[SCollection[V]], index: Int)
+case class ByIndex[V <: SType](input: Value[SCollection[V]], index: Value[SInt.type])
   extends Transformer[SCollection[V], V] with NotReadyValue[V] with Rewritable {
   override val opCode: OpCode = OpCodes.ByIndexCode
 
-  def tpe = input.tpe.elemType
+  override def tpe = input.tpe.elemType
+  override def arity = 3
+  override def deconstruct = immutable.Seq[Any](input, index, tpe)
 
-  def arity = 3
-
-  def deconstruct = immutable.Seq[Any](input, index, tpe)
-
-  def reconstruct(cs: immutable.Seq[Any]) = cs match {
-    case Seq(input: Value[SCollection[V]]@unchecked, index: Int, _) => ByIndex[V](input, index)
-    case _ => illegalArgs("ByIndex", "(Value[SCollection[V]], index: Int)(tpe: V)", cs)
+  override def reconstruct(cs: immutable.Seq[Any]): ByIndex[V] = cs match {
+    case Seq(input: Value[SCollection[V]]@unchecked, index: Value[SInt.type]@unchecked, _) =>
+      ByIndex[V](input, index)
+    case _ => illegalArgs("ByIndex", "(Value[SCollection[V]], index: Value[SInt.type])(tpe: V)", cs)
   }
 
-  override def function(input: EvaluatedValue[SCollection[V]]) =
-    input.items.apply(index)
+  override def function(input: EvaluatedValue[SCollection[V]]): Value[V] =
+    input.items.apply(index.asInstanceOf[EvaluatedValue[SInt.type]].value.toInt)
```

Comment:
I see no any guarantee why index would be evaluated at the moment of function() being calling. I guess transformationReady() should also be redefined. And tests for index evaluation should be done.

Reply:
Oops, I missed that big time. Thanks. On it.

Reply:
@kushti Fixed.